lib/src/optimism/batcher_db.rs

```diff
@@ -0,0 +1,149 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use anyhow::{bail, Result};
+use serde::{Deserialize, Serialize};
+use zeth_primitives::{
+    block::Header,
+    receipt::Receipt,
+    transactions::{
+        ethereum::EthereumTxEssence, optimism::OptimismTxEssence, Transaction, TxEssence,
+    },
+    trie::MptNode,
+    RlpBytes,
+};
+
+use crate::optimism::{config::OPTIMISM_CHAIN_SPEC, deposits, system_config};
+
+/// Input for extracting deposits.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct BlockInput<E: TxEssence> {
+    /// Header of the block.
+    pub block_header: Header,
+    /// Transactions of the block.
+    pub transactions: Vec<Transaction<E>>,
+    /// Transaction receipts of the block or `None` if not required.
+    pub receipts: Option<Vec<Receipt>>,
+}
+
+pub trait BatcherDb {
+    fn get_full_op_block(&mut self, block_no: u64) -> Result<BlockInput<OptimismTxEssence>>;
+    fn get_op_block_header(&mut self, block_no: u64) -> Result<Header>;
+    fn get_full_eth_block(&mut self, block_no: u64) -> Result<BlockInput<EthereumTxEssence>>;
+    fn get_eth_block_header(&mut self, block_no: u64) -> Result<Header>;
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct MemDb {
+    pub full_op_block: HashMap<u64, BlockInput<OptimismTxEssence>>,
+    pub op_block_header: HashMap<u64, Header>,
+    pub full_eth_block: HashMap<u64, BlockInput<EthereumTxEssence>>,
+    pub eth_block_header: HashMap<u64, Header>,
+}
+
+impl MemDb {
+    pub fn new() -> Self {
+        MemDb {
+            full_op_block: HashMap::new(),
+            op_block_header: HashMap::new(),
+            full_eth_block: HashMap::new(),
+            eth_block_header: HashMap::new(),
+        }
+    }
+}
+
+impl Default for MemDb {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BatcherDb for MemDb {
+    fn get_full_op_block(&mut self, block_no: u64) -> Result<BlockInput<OptimismTxEssence>> {
+        let op_block = self.full_op_block.remove(&block_no).unwrap();
+        assert_eq!(block_no, op_block.block_header.number);
+
+        // Validate tx list
+        {
+            let mut tx_trie = MptNode::default();
+            for (tx_no, tx) in op_block.transactions.iter().enumerate() {
+                let trie_key = tx_no.to_rlp();
+                tx_trie.insert_rlp(&trie_key, tx)?;
+            }
+            if tx_trie.hash() != op_block.block_header.transactions_root {
+                bail!("Invalid op block transaction data!")
+            }
+        }
+
```

Comment:
We don't validate the receipts for the OP blocks, so I guess we should instead just make sure that the receipts field in the BlockInput is indeed empty.

Reply:
Agreed. Done